core/packages/contracts/src/NativeTokens.sol

```diff
@@ -0,0 +1,149 @@
+contract NativeTokens is AccessControl {
+    /// @dev Describes the type of message.
+    enum Action {
+        Unlock
+    }
+
+    /// @dev Message format.
+    struct Message {
+        /// @dev The action type.
+        Action action;
+        /// @dev The message payload.
+        bytes payload;
+    }
+
+    /// @dev Unlock payload format.
+    struct UnlockPayload {
+        /// @dev The ERC20 token to unlock.
+        address token;
+        /// @dev The destination address that will receive unlocked funds.
+        address recipient;
+        /// @dev The amount to unlock.
+        uint128 amount;
+    }
+
+    /// @dev Emitted once the funds are locked and a message is successfully queued.
+    event Locked(bytes recipient, address token, uint128 amount);
+
+    /// @dev Emitted once the funds are unlocked.
+    event Unlocked(address recipient, address token, uint128 amount);
+
+    /// @dev Emitted after enqueueing a a create token message to substrate.
+    event Created(address token);
+
+    /// @dev Set a new outbound channel.
+    event OutboundChannelUpdated(address newOutBoundChannel);
+
+    /* State */
+
+    bytes32 public constant ADMIN_ROLE = keccak256("ADMIN_ROLE");
+    bytes32 public constant SENDER_ROLE = keccak256("SENDER_ROLE");
+
+    bytes32 public immutable peerID;
+    bytes public peer;
+
+    TokenVault public immutable vault;
+    IOutboundChannel public outboundChannel;
+
+    uint256 public createTokenFee;
+
+    /* Errors */
+
+    error InvalidAmount();
+    error Unauthorized();
+    error NoFundsforCreateToken();
+
+    constructor(
+        TokenVault _vault,
+        IOutboundChannel _outboundChannel,
+        bytes memory _peer,
+        uint256 _createTokenFee
+    ) {
+        _grantRole(ADMIN_ROLE, msg.sender);
+        _setRoleAdmin(SENDER_ROLE, ADMIN_ROLE);
+        vault = _vault;
+        outboundChannel = _outboundChannel;
+        peer = _peer;
+        peerID = keccak256(_peer);
+        createTokenFee = _createTokenFee;
+    }
+
+    /// @dev Locks an amount of ERC20 Tokens in the vault and enqueues a mint message.
+    /// Requires the allowance to be set on the ERC20 token where the spender is the vault.
+    /// @param token The token to lock.
+    /// @param recipient The recipient on the substrate side.
+    /// @param amount The amount to lock.
+    function lock(address token, bytes calldata recipient, uint128 amount) external payable {
+        if (amount == 0) {
+            revert InvalidAmount();
+        }
+
+        vault.deposit(msg.sender, token, amount);
+
+        bytes memory payload = NativeTokensTypes.Mint(peer, token, recipient, amount);
+        outboundChannel.submit{ value: msg.value }(peer, payload);
+
+        emit Locked(recipient, token, amount);
+    }
+
+    /// @dev Enqueues a create native token message to substrate.
+    /// @param token The ERC20 token address.
+    function create(address token) external payable {
```

Comment:
Is `create` here can execute by anyone? is it better we add a `CREATE_ROLE` for it?

Reply:
Yeah, that's by design. Its permissionless. So anyone can call this to register a new token on Statemint.

This is to prevent us from being overwhelmed by governance requests to register new ERC20 tokens with the bridge.

We do make sure to charge a high fee for calling `create`, to avoid spamming attacks.